<a href="https://colab.research.google.com/github/Chenghan-Sun/MoA-Prediction/blob/master/notebooks/MoA_NODE.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## 2.3 Experiment of NEURAL OBLIVIOUS DECISION ENSEMBLES (NODE) on MoA dataset

In [1]:
# !git clone https://github.com/Qwicen/node.git
# !pip install -r node/requirements.txt
# !pip install qhoptim

# only need this one
! pip install iterative-stratification

In [2]:
import sys, os
import gc
import datetime
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from typing import Optional
from tqdm.notebook import tqdm
from time import time
import warnings
warnings.filterwarnings("ignore")

import tensorflow as tf
from tensorflow.keras.callbacks import ReduceLROnPlateau, EarlyStopping
import tensorflow_addons as tfa
import tensorflow_probability as tfp

# from sklearn.model_selection import KFold
from sklearn.metrics import log_loss
from iterstrat.ml_stratifiers import MultilabelStratifiedKFold

Run the following cell if you want to use GDrive (file stored under `/content/drive/MyDrive/Colab Notebooks/MoA-Prediction/moa_data/`.

In [3]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [4]:
# path in GDrive
# data_path = "/content/drive/MyDrive/Colab Notebooks/MoA/lish-moa/"

# path
data_path = "/content/drive/MyDrive/Colab Notebooks/MoA-Prediction/moa_data/"
data_file_list = ["train_features.csv", "test_features.csv", "train_targets_scored.csv", \
                  "train_targets_nonscored.csv"]

# load data
data_train = pd.read_csv(data_path + data_file_list[0])
data_test = pd.read_csv(data_path + data_file_list[1])
train_targets_scored = pd.read_csv(data_path + data_file_list[2])
train_targets_nonscored = pd.read_csv(data_path + data_file_list[3])

# data info
print(f'Training features file: {data_train.shape[0]} rows; {data_train.shape[1]} columns')
print(f'Testing features file: {data_test.shape[0]} rows; {data_test.shape[1]} columns')
print(f'Training targets scored file: {train_targets_scored.shape[0]} rows; \
{train_targets_scored.shape[1]} columns')
print(f'Training targets nonscored file: {train_targets_nonscored.shape[0]} rows; \
{train_targets_nonscored.shape[1]} columns')

Training features file: 23814 rows; 876 columns
Testing features file: 3982 rows; 876 columns
Training targets scored file: 23814 rows; 207 columns
Training targets nonscored file: 23814 rows; 403 columns


In [5]:
data_train.head(3)

sig_id cp_type  cp_time cp_dose  ...    c-96    c-97    c-98    c-99
0  id_000644bb2  trt_cp       24      D1  ... -0.3981  0.2139  0.3801  0.4176
1  id_000779bfc  trt_cp       72      D1  ...  0.1522  0.1241  0.6077  0.7371
2  id_000a6266a  trt_cp       48      D1  ... -0.6417 -0.2187 -1.4080  0.6931

[3 rows x 876 columns]

In [6]:
data_test.head(3)

sig_id      cp_type  cp_time cp_dose  ...    c-96    c-97    c-98    c-99
0  id_0004d9e33       trt_cp       24      D1  ...  0.0210 -0.0502  0.1510 -0.7750
1  id_001897cda       trt_cp       72      D1  ...  0.0359 -0.4764 -1.3810 -0.7300
2  id_002429b5b  ctl_vehicle       24      D1  ...  0.8662  1.0160  0.4924 -0.1942

[3 rows x 876 columns]

In [7]:
train_targets_scored.sample(3)

sig_id  ...  wnt_inhibitor
12771  id_898db4ee1  ...              0
22554  id_f23c45128  ...              0
9646   id_67f174dac  ...              0

[3 rows x 207 columns]

In [8]:
data_train.cp_type.value_counts(normalize=True)

trt_cp         0.921643
ctl_vehicle    0.078357
Name: cp_type, dtype: float64

In [9]:
control_group = data_train.loc[data_train.cp_type == 'ctl_vehicle']
control_group['sig_id'].count()

1866

**Summary**:

7.8357% samples in the original dataset were treated as the control perturbation => 1866 samples out of total sample pool of 23814. In the dataset, cp_type indicates samples treated with a compound (cp_vehicle) or with a control perturbation (ctrl_vehicle); Since **control perturbations have no MoAs**, samples (sig_id) with this cp_type will be droped from training. 

### Pre-processing

In [10]:
# define helper functions data pre-processing

def df_pre_processing(raw_df, type='training', verbose=True):
    # expand features 2 non-numerical features 'cp_type', 'cp_dose' to 4 dummy 
    # features based on categorical values 
    processed_df = pd.concat([raw_df, pd.get_dummies(raw_df['cp_dose'], prefix='cp_dose')], axis=1)
    processed_df = pd.concat([processed_df, pd.get_dummies(raw_df['cp_type'], \
                                                                           prefix='cp_type')], axis=1)

    # drop the three original features
    processed_df = processed_df.drop(['cp_type', 'cp_dose'], axis=1)

    # removed the samples with wrong cp_type -- removed 1866 samples
    processed_df = processed_df.loc[processed_df['cp_type_trt_cp']==1].reset_index(drop=True)
    
    # drop the original sig_id column
    processed_df = processed_df.drop(columns='sig_id')

    # show shape of processed df
    if verbose:
        print(f"Processed {type} dataset shape = {processed_df.shape}.")
        
    return processed_df

# apply on both training and test dataset
data_train_processed = df_pre_processing(raw_df=data_train)
data_test_processed = df_pre_processing(raw_df=data_test, type='test')

data_train_processed.head(3)

Processed training dataset shape = (21948, 877).
Processed test dataset shape = (3624, 877).


cp_time     g-0     g-1  ...  cp_dose_D2  cp_type_ctl_vehicle  cp_type_trt_cp
0       24  1.0620  0.5577  ...           0                    0               1
1       72  0.0743  0.4087  ...           0                    0               1
2       48  0.6280  0.5817  ...           0                    0               1

[3 rows x 877 columns]

In [11]:
# define helper functions for traget multi-binary-label classification metric

def metric(y_true, y_pred, df_train_targets):
    metrics = []
    for _target in df_train_targets.columns:
        metrics.append(log_loss(y_true.loc[:, _target], y_pred.loc[:, _target].astype(float), labels=[0,1]))
    return np.mean(metrics)

def get_average_metrics_out_of_cv_folds(history_list):
    """ Derive total number of folds and total number epochs"""
    total_n_folds = len(history_list)
    total_epochs = len(hist_total_list[0].history['loss'])
    mean_loss = np.zeros([total_n_folds,total_epochs])
    mean_val_loss = np.zeros([total_n_folds,total_epochs])
    
    # Put loss and validation loss in each fold
    for i in range(total_n_folds):
        mean_loss[i,:] = history_list[i].history['loss']
        mean_val_loss[i,:] = history_list[i].history['val_loss']

    # Get average loss and validation loss
    mean_loss = np.mean(mean_loss,axis=0)
    mean_val_loss  = np.mean(mean_val_loss ,axis=0)
    return mean_loss, mean_val_loss

In [12]:
# removed the samples with wrong cp_type on the training targets scored dataset
train_targets_w_id = train_targets_scored.loc[data_train['cp_type']=='trt_cp'].reset_index(drop=True)

# drop the original sig_id column
train_targets = train_targets_w_id.drop(columns='sig_id')

print(f"The processed training targets scored data shape = {train_targets.shape}.")

The processed training targets scored data shape = (21948, 206).


### Implementation of NODE based on tf.keras framework 

In [13]:
@tf.function
def sparsemoid(inputs: tf.Tensor):
    return tf.clip_by_value(0.5 * inputs + 0.5, 0., 1.)

@tf.function
def identity(x: tf.Tensor):
    return x

class ODST(tf.keras.layers.Layer):
    def __init__(self, n_trees: int = 3, depth: int = 4, units: int = 1, threshold_init_beta: float = 1.):
        super(ODST, self).__init__()
        self.initialized = False
        self.n_trees = n_trees
        self.depth = depth
        self.units = units
        self.threshold_init_beta = threshold_init_beta
    
    def build(self, input_shape: tf.TensorShape):
        feature_selection_logits_init = tf.zeros_initializer()
        self.feature_selection_logits = tf.Variable(initial_value=feature_selection_logits_init(shape=(input_shape[-1], self.n_trees, self.depth), dtype='float32'),
                                 trainable=True)        
        
        feature_thresholds_init = tf.zeros_initializer()
        self.feature_thresholds = tf.Variable(initial_value=feature_thresholds_init(shape=(self.n_trees, self.depth), dtype='float32'),
                                 trainable=True)
        
        log_temperatures_init = tf.ones_initializer()
        self.log_temperatures = tf.Variable(initial_value=log_temperatures_init(shape=(self.n_trees, self.depth), dtype='float32'),
                                 trainable=True)
        
        indices = tf.keras.backend.arange(0, 2 ** self.depth, 1)
        offsets = 2 ** tf.keras.backend.arange(0, self.depth, 1)
        bin_codes = (tf.reshape(indices, (1, -1)) // tf.reshape(offsets, (-1, 1)) % 2)
        bin_codes_1hot = tf.stack([bin_codes, 1 - bin_codes], axis=-1)
        self.bin_codes_1hot = tf.Variable(initial_value=tf.cast(bin_codes_1hot, 'float32'), 
                                          trainable=False)
        
        response_init = tf.ones_initializer()
        self.response = tf.Variable(initial_value=response_init(shape=(self.n_trees, self.units, 2**self.depth), dtype='float32'), 
                                    trainable=True)
                
    def initialize(self, inputs):        
        feature_values = self.feature_values(inputs)
        
        # intialize feature_thresholds
        percentiles_q = (100 * tfp.distributions.Beta(self.threshold_init_beta, 
                                                      self.threshold_init_beta)
                         .sample([self.n_trees * self.depth]))
        flattened_feature_values = tf.map_fn(tf.keras.backend.flatten, feature_values)
        init_feature_thresholds = tf.linalg.diag_part(tfp.stats.percentile(flattened_feature_values, percentiles_q, axis=0))
        
        self.feature_thresholds.assign(tf.reshape(init_feature_thresholds, self.feature_thresholds.shape))
        
        # intialize log_temperatures
        self.log_temperatures.assign(tfp.stats.percentile(tf.math.abs(feature_values - self.feature_thresholds), 50, axis=0))
        
    def feature_values(self, inputs: tf.Tensor, training: bool = None):
        feature_selectors = tfa.activations.sparsemax(self.feature_selection_logits)
        # ^--[in_features, n_trees, depth]

        feature_values = tf.einsum('bi,ind->bnd', inputs, feature_selectors)
        # ^--[batch_size, n_trees, depth]
        
        return feature_values
        
    def call(self, inputs: tf.Tensor, training: bool = None):
        if not self.initialized:
            self.initialize(inputs)
            self.initialized = True
            
        feature_values = self.feature_values(inputs)
        
        threshold_logits = (feature_values - self.feature_thresholds) * tf.math.exp(-self.log_temperatures)

        threshold_logits = tf.stack([-threshold_logits, threshold_logits], axis=-1)
        # ^--[batch_size, n_trees, depth, 2]

        bins = sparsemoid(threshold_logits)
        # ^--[batch_size, n_trees, depth, 2], approximately binary

        bin_matches = tf.einsum('btds,dcs->btdc', bins, self.bin_codes_1hot)
        # ^--[batch_size, n_trees, depth, 2 ** depth]

        response_weights = tf.math.reduce_prod(bin_matches, axis=-2)
        # ^-- [batch_size, n_trees, 2 ** depth]

        response = tf.einsum('bnd,ncd->bnc', response_weights, self.response)
        # ^-- [batch_size, n_trees, units]
        
        return tf.reduce_sum(response, axis=1)
        
class NODE(tf.keras.Model):
    def __init__(self, units: int = 1, n_layers: int = 1, output_dim = 1, dropout_rate = 0.1, link: tf.function = tf.identity, n_trees: int = 3, depth: int = 4, threshold_init_beta: float = 1., feature_column: Optional[tf.keras.layers.DenseFeatures] = None):
        super(NODE, self).__init__()
        self.units = units
        self.n_layers = n_layers
        self.n_trees = n_trees
        self.depth = depth
        self.units = units
        self.threshold_init_beta = threshold_init_beta
        self.feature_column = feature_column
        self.dropout_rate = dropout_rate
        self.output_dim = output_dim
        
        if feature_column is None:
            self.feature = tf.keras.layers.Lambda(identity)
        else:
            self.feature = feature_column
        
        self.bn = [tf.keras.layers.BatchNormalization() for _ in range(n_layers + 1)]
        self.dropout = [tf.keras.layers.Dropout(self.dropout_rate) for _ in range(n_layers + 1)]
        self.ensemble = [ODST(n_trees = n_trees,
                              depth = depth,
                              units = units,
                              threshold_init_beta = threshold_init_beta) 
                         for _ in range(n_layers)]
        
        self.last_layer = tf.keras.layers.Dense(self.output_dim)
        
        self.link = link
        
    def call(self, inputs, training=None):
        X = self.feature(inputs)
        X = self.bn[0](X, training=training)
        X = self.dropout[0](X, training=training)
        
        for i, tree in enumerate(self.ensemble):
            H = tree(X)
            X = tf.concat([X, H], axis=1)
            X = self.bn[i + 1](X, training=training)
            X = self.dropout[i + 1](X, training=training)
            
        return self.link(self.last_layer(X))

In [14]:
def compile_NODE(n_layers, units, output_dim, dropout_rate, depth, n_trees, link, learning_rate):
    """ create 
    """ 
    node = NODE(n_layers=n_layers, units=units, output_dim=output_dim, dropout_rate=dropout_rate, 
                depth=depth, n_trees=n_trees, link=tf.keras.activations.sigmoid)
    
    node.compile(optimizer = tf.keras.optimizers.Adam(learning_rate=learning_rate), 
                 loss = 'binary_crossentropy')
    return node

## Train model

In [15]:
# new version of NODE training code
# integrated with K-Fold CV

hist_total_list = []
df_residual = train_targets.copy()
df_residual.loc[:, train_targets.columns] = 0  # reset the residual matrix

# set global parameters 
N_EXPTS = 3
N_FOLDS = 5

for seed in range(N_EXPTS):
    start_time_seed = time()
    tf.keras.backend.clear_session()  # clear previous tf session 
    tf.random.set_seed(seed)
    # Stratify the Multilabel K-Folds cross-validator 
    mlsk = MultilabelStratifiedKFold(n_splits=N_FOLDS, random_state=42, shuffle=True)
    mean_loss = 0

    # split training targets based on number of folds 
    for fold, (tr, val) in enumerate(mlsk.split(train_targets, train_targets)):
        start_time_fold = time()  # track time per fold
        x_tr, x_val = data_train_processed.values[tr][:,:], data_train_processed.values[val][:,:]
        y_tr, y_val = train_targets.values[tr][:,:], train_targets.values[val][:,:]
        
        # NODE model instance
        node_model = compile_NODE(n_layers=3, dropout_rate=0.1, depth=3, 
                                  n_trees=3, link=tf.keras.activations.sigmoid, 
                                  learning_rate=1e-3, units=32, output_dim=206)
        # adaptive alpha, monitor the learning rate and reduce the rate after 3 epochs
        rlr = ReduceLROnPlateau(monitor='val_loss', factor=0.1, patience=3, 
                                verbose=0, min_delta=1e-4, mode='min')
        # stop training when val_loss has stopped improving
        es = EarlyStopping(monitor='val_loss', min_delta=1e-4, patience=5, 
                           mode='min', restore_best_weights=True, verbose=0)
        
        # start fitting NODE model
        history = node_model.fit(x_tr, y_tr, validation_data = (x_val, y_val), 
                                 epochs=30, batch_size=128, callbacks=[rlr, es], verbose=0)
        
        hist_total_list.append(history)
        hist = pd.DataFrame(history.history)
        fold_loss = hist['val_loss'].min()
        mean_loss += fold_loss / N_FOLDS

        # save predicted prob in residual df
        val_predict = node_model.predict(data_train_processed.values[val][:,:])
        df_residual.loc[val, train_targets.columns] += val_predict / N_EXPTS
        print(f'Loss @ Seed {seed}, Fold {fold} = {fold_loss}; time usage = \
        {str(datetime.timedelta(seconds=time()-start_time_fold))[2:7]}')
        
    print(f'Mean loss @ Seed {seed} = {mean_loss}; time usage = \
    {str(datetime.timedelta(seconds=time()-start_time_seed))[2:7]}')

Instructions for updating:
This op will be removed after the deprecation date. Please switch to tf.sets.difference().
[00:54]         Loss @ Seed 0, Fold 0 = 0.017161965370178223
[00:48]         Loss @ Seed 0, Fold 1 = 0.017339009791612625
[01:00]         Loss @ Seed 0, Fold 2 = 0.016740186139941216
[00:51]         Loss @ Seed 0, Fold 3 = 0.017274856567382812
[00:45]         Loss @ Seed 0, Fold 4 = 0.017443062737584114
[04:22]     Mean loss @ Seed 0 = 0.017191816121339798
[00:54]         Loss @ Seed 1, Fold 0 = 0.01719517819583416
[00:45]         Loss @ Seed 1, Fold 1 = 0.01736970990896225
[00:54]         Loss @ Seed 1, Fold 2 = 0.01676858589053154
[00:59]         Loss @ Seed 1, Fold 3 = 0.017223818227648735
[00:53]         Loss @ Seed 1, Fold 4 = 0.017309993505477905
[04:30]     Mean loss @ Seed 1 = 0.017173457145690917
[00:51]         Loss @ Seed 2, Fold 0 = 0.01720501482486725
[00:57]         Loss @ Seed 2, Fold 1 = 0.01735789328813553
[00:51]         Loss @ Seed 2, Fold 2 = 0.01675

In [16]:
df_residual.loc[1, train_targets.columns]

5-alpha_reductase_inhibitor              0.001043
11-beta-hsd1_inhibitor                   0.000495
acat_inhibitor                           0.001042
acetylcholine_receptor_agonist           0.017400
acetylcholine_receptor_antagonist        0.014149
                                           ...   
ubiquitin_specific_protease_inhibitor    0.000152
vegfr_inhibitor                          0.002796
vitamin_b                                0.001536
vitamin_d_receptor_agonist               0.006526
wnt_inhibitor                            0.003092
Name: 1, Length: 206, dtype: float64

Note on expt:

- performance baseline: 0.017787196522898216
- sigmoid to softmax: 
    - no change 
- n_tree
- n_layer
    - change from 3 to 5: 0.017573646823844536
- depth

- n_layer=5, n_tree=5, dropout_rate=0.05, depth=6 -> 0.017731386552524063

- n_layer=3, n_tree=5, dropout_rate=0.1, depth=6 -> 0.01814203681110645

- n_layer=3, n_tree=3, dropout_rate=0.1, depth=6 -> 0.017606996746624328

- n_layer=3, n_tree=3, dropout_rate=0.1, depth=9 -> 0.017835741417378655

In [17]:
print(df_residual.shape)
print(train_targets.shape)

print(f'NODE OOF Metric: {metric(train_targets, df_residual, train_targets)}')

(21948, 206)
(21948, 206)
NODE OOF Metric: 0.016978945140112986


In [22]:
# save important results 

np.save('hist_total_list.npy', hist_total_list)

PicklingError: ignored

### Plot Loss and Accuracy vs. Epochs

In [20]:
# Get average loss and validation loss out of all folds
mean_loss, mean_val_loss = get_average_metrics_out_of_cv_folds(hist_total_list)

# Plot Loss and Accuracy vs. Epochs
plt.figure(figsize=(12, 8))
plt.plot(np.arange(1,len(mean_loss)+1,1),np.array(mean_loss),'-s',lw=4,label=r'$loss_{training}$',c='r')
plt.plot(np.arange(1,len(mean_val_loss )+1,1),np.array(mean_val_loss ),'--o',lw=4,label=r'$loss_{validation}$',c='r')
plt.ylabel('Loss',fontsize = 25,fontweight='bold')
plt.xlabel("Epochs",fontsize = 25,fontweight='bold')
plt.legend(loc=1,frameon=False)
plt.show()

ValueError: ignored